File: analysis/python/cpoe.py
"""M4 (Eager & Erickson, Ch.5) — Completion % Over Expected, Python parity of cpoe.R.

statsmodels GLM(Binomial) mirrors R's glm(family=binomial) (both IRLS) → coefficients agree to
tolerance. Same feed, features, and filtering as the R version.

Reads data/gold/plays_model.csv; writes coef__cpoe__py / metrics__cpoe__py / pred__cpoe__py.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.api as sm

from cfb_analytics.config import REPO_ROOT

GOLD = REPO_ROOT / "data" / "gold"
RESULTS = REPO_ROOT / "data" / "results"
FEATURES = ["down", "distance", "yards_to_goal", "offense_score_diff_start"]
_TERM = {"Intercept": "(Intercept)"}


def main() -> None:
    plays = pd.read_csv(GOLD / "plays_model.csv")
    pas = plays[(plays.is_pass_attempt == 1) & plays.is_completion.notna()
                & plays.down.notna() & plays.distance.notna()
                & plays.yards_to_goal.notna()
                & plays.offense_score_diff_start.notna()].copy()

    res = smf.glm("is_completion ~ " + " + ".join(FEATURES),
                  data=pas, family=sm.families.Binomial()).fit()

    prob = res.predict(pas).to_numpy()
    y = pas["is_completion"].to_numpy()
    cpoe = y - prob

    brier = float(np.mean((y - prob) ** 2))
    base_rate = float(y.mean())
    brier_base = float(np.mean((y - base_rate) ** 2))
    p = np.clip(prob, 1e-15, 1 - 1e-15)
    log_loss = float(-np.mean(y * np.log(p) + (1 - y) * np.log(1 - p)))

    coef = pd.DataFrame({
        "model": "cpoe",
        "term": [_TERM.get(t, t) for t in res.params.index],
        "estimate": res.params.values,
        "std_error": res.bse.values,
        "statistic": res.tvalues.values,
        "p_value": res.pvalues.values,
        "odds_ratio": np.exp(res.params.values),
        "language": "py",
    })
    metrics = pd.DataFrame([
        dict(model="cpoe", metric="brier", value=brier),
        dict(model="cpoe", metric="brier_baseline", value=brier_base),
        dict(model="cpoe", metric="log_loss", value=log_loss),
        dict(model="cpoe", metric="completion_rate", value=base_rate),
        dict(model="cpoe", metric="n_obs", value=float(len(pas))),
    ])
    metrics["language"] = "py"
    pred = pd.DataFrame({
        "play_key": pas["play_key"].values,
        "completion_prob": prob,
        "cpoe": cpoe,
        "language": "py",
    })

    RESULTS.mkdir(parents=True, exist_ok=True)
    coef.to_csv(RESULTS / "coef__cpoe__py.csv", index=False)
    metrics.to_csv(RESULTS / "metrics__cpoe__py.csv", index=False)
    pred.to_csv(RESULTS / "pred__cpoe__py.csv", index=False)
    print(f"[cpoe/py] Brier={brier:.4f} (baseline {brier_base:.4f}), "
          f"logloss={log_loss:.4f}, {len(pas)} attempts")


if __name__ == "__main__":
    main()

File: analysis/R/cpoe.R
# M4 (Eager & Erickson, Ch.5) — Completion Percentage Over Expected in R.
# GLM / logistic regression: model the PROBABILITY a pass is completed from game situation;
# the residual (actual completion - predicted prob) is QB/offense skill = CPOE. We have no
# air-yards in the feed, so situation = down + distance + yards_to_goal + score diff.
#   glm(is_completion ~ ..., family = binomial)
# Odds ratios (exp(coef)) are reported as the book emphasises.
#
# Reads:  data/gold/plays_model.csv
# Writes: data/results/coef__cpoe__r.csv, metrics__cpoe__r.csv, pred__cpoe__r.csv

source("analysis/R/util_io.R")
suppressPackageStartupMessages({ library(dplyr); library(broom) })
set.seed(42)

FEATURES <- c("down", "distance", "yards_to_goal", "offense_score_diff_start")

plays <- cfb_read_gold("plays_model")

pass <- plays %>%
  filter(is_pass_attempt == 1, !is.na(is_completion),
         !is.na(down), !is.na(distance), !is.na(yards_to_goal),
         !is.na(offense_score_diff_start))

f <- as.formula(paste("is_completion ~", paste(FEATURES, collapse = " + ")))
m <- glm(f, data = pass, family = binomial())

prob <- predict(m, pass, type = "response")
cpoe <- pass$is_completion - prob

# Brier score (mean squared error of the probability) vs the base-rate baseline it must beat.
brier <- mean((pass$is_completion - prob)^2)
base_rate <- mean(pass$is_completion)
brier_base <- mean((pass$is_completion - base_rate)^2)
# log-loss
eps <- 1e-15; p <- pmin(pmax(prob, eps), 1 - eps)
log_loss <- -mean(pass$is_completion * log(p) + (1 - pass$is_completion) * log(1 - p))

coef <- tidy(m) %>%
  transmute(model = "cpoe", term, estimate, std_error = std.error,
            statistic, p_value = p.value,
            odds_ratio = exp(estimate), language = "r")

metrics <- data.frame(
  model = "cpoe",
  metric = c("brier", "brier_baseline", "log_loss", "completion_rate", "n_obs"),
  value = c(brier, brier_base, log_loss, base_rate, nrow(pass)),
  language = "r")

pred <- data.frame(
  play_key = pass$play_key,
  completion_prob = prob,
  cpoe = cpoe,
  language = "r")

cfb_write_result("coef__cpoe__r", coef)
cfb_write_result("metrics__cpoe__r", metrics)
cfb_write_result("pred__cpoe__r", pred)
cat(sprintf("[cpoe/R] Brier=%.4f (baseline %.4f), logloss=%.4f, %d attempts\n",
            brier, brier_base, log_loss, nrow(pass)))
